libs/pq_proto/src/framed.rs

```diff
@@ -0,0 +1,265 @@
+//! Provides `Framed` -- writing/flushing and reading Postgres messages to/from
+//! the async stream based on (and buffered with) BytesMut. All functions are
+//! cancellation safe.
+//!
+//! It is similar to what tokio_util::codec::Framed with appropriate codec
```

Comment:
Yes, the only supposed way to use split Framed is split on Stream/Sink:
https://docs.rs/tokio-util/latest/tokio_util/codec/struct.Framed.html#method.from_parts

See split.rs and bilock.rs in futures-util.

Funnily, there is even a crate doing bilock with allocation:
https://github.com/paritytech/bilock

Reply:
The question still remains. Why we cannot use existing `tokio_util::codec::Framed`?

Reply:
Because I don't like allocation in the hot path of reading/writing each message in split case (which is main for safekeepers). Additionally, though less important, as this comment says having same error for all io framed functions is better avoided. Also we don't really need Stream/Sink abstractions here (apart from mental/potential perf overhead it creates friction like returning `Option<Result<...>>` on next message, while `Result<Option<...>>` is easier to handle).

Reply:
Can you please point to the place where these allocations happen?

>  as this comment says having same error for all io framed functions is better avoided

> Will an enum with different error variants be a decent way to handle that?

I'm not sure I understand which methods are problematic. Can you elaborate on that?

> returning Option<Result<...>> on next message, while Result<Option<...>> is easier to handle).

See [Option::transpose](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.transpose). Will it help? 

My concern is the code that in some places duplicates code from library (futures, tokio), and I still do not understand why do we need to invent this ourselves.

Reply:
https://docs.rs/futures-util/latest/src/futures_util/lock/bilock.rs.html#107

In the final version of the patch in safekeeper we read-write only from the same task/thread (currently; in theory this might also change), so on the first glance this might be not an issue . I still don't think we'd have sensible simplicity win with tokio_util Framed though; currently all related logic is last ~50 LOC in `framed.rs`. tokio_util Framed itself is more lenghty and complicated in ways we don't need. Finally, @funbringer expressed eager desire to have .await's in protocol parsing functions, that's not compatible with tokio_util Framed as well.

> I'm not sure I understand which methods are problematic.

read_message, write_message, flush all return different errors reflecting what they do. Having function declared as returning variants it never actually returns makes code harder to read.

> See [Option::transpose](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.transpose). 

Nice, I found [thread](https://users.rust-lang.org/t/convenience-method-for-flipping-option-result-to-result-option/13695/24) when googled that, but haven't scrolled to the bottom. This is minor though.

Reply:
> https://docs.rs/futures-util/latest/src/futures_util/lock/bilock.rs.html#107

Ok, makes sense, please add this to the comment if we will keep this code. From the other side, in tokio instead of keeping wakers around it spins when trying to acquire the lock https://docs.rs/tokio/latest/src/tokio/io/split.rs.html#145. Not sure if this is a problem though.

> I still don't think we'd have sensible simplicity win with tokio_util Framed though; currently all related logic is last ~50 LOC in framed.rs. tokio_util Framed itself is more lenghty and complicated in ways we don't need.

I wont insist. To me tokio's framed can allow for better composition with other tokio related utilities. It may be reasonable to keep the scope smaller. Lets see if we can do better with @funbringer's approach

Reply:
added that link to the comment